src/vmm_lwt.ml

```diff
@@ -173,3 +173,75 @@ let write_wire s wire =
   Bytes.set_int32_be dlen 0 (Int32.of_int (String.length data)) ;
   let buf = Bytes.cat dlen (Bytes.unsafe_of_string data) in
   write_raw s buf
+
+let compress_stream ~level input =
+  let w = De.Lz77.make_window ~bits:15 in
+  let q = De.Queue.create 0x1000 in
+  let zl = Zl.Def.encoder `Manual `Manual ~q ~w ~level in
+  let i = Bigstringaf.create De.io_buffer_size in
+  let o = Bigstringaf.create De.io_buffer_size in
+  let r, stream = Lwt_stream.create_bounded 2 in
+  let rec loop aw zl =
+    match Zl.Def.encode zl with
+    | `Flush zl ->
+      if aw then begin
+        let len = Bigstringaf.length o - Zl.Def.dst_rem zl in
+        stream#push (Bigstringaf.substring o ~off:0 ~len) >>= fun () ->
+        loop aw (Zl.Def.dst zl o 0 (Bigstringaf.length o))
+      end else
+        loop aw (Zl.Def.dst zl o 0 (Bigstringaf.length o))
+    | `End zl ->
+      let len = Bigstringaf.length o - Zl.Def.dst_rem zl in
+      (if len > 0 then
+         stream#push (Bigstringaf.substring o ~off:0 ~len)
+       else Lwt.return_unit) >|= fun () ->
+      stream#close
+    | `Await zl ->
+      Lwt_stream.get input >>= function
+      | Some data ->
+        (* XXX(reynir): what if [data] doesn't fit?! *)
```

Comment:
Another way around to solve this situation is to make an inner loop which tries to consume `(data, off (* initialy to 0 *), len (* initally to String.length data *))` and execute the loop until `len = 0`.